lib/middlewares/object_authorization.ex

```
@@ -114,6 +122,13 @@ defmodule Rajska.ObjectAuthorization do
     authorize(schema_node, selections ++ tail, resolution)
   end
 
+  defp find_associations(
+    [%Absinthe.Blueprint.Document.Fragment.Spread{} | tail],
```

I just did a PR for this and only now realized you already did something similar. But I think this implementation is not the best because we simply ignore the fragments, which could contain further associations. 

What do you think about the implementation in https://github.com/jungsoft/rajska/pull/29?

---

Hey 👋 , good catch, however, I think I added line 66, which should project the fields down. If you look at the docs for [`Absinthe.Resolution.project/2`](https://github.com/absinthe-graphql/absinthe/blob/v1.5.2/lib/absinthe/resolution/projector.ex#L6-L13), they mention it's non-trivial. The fragments might not be simple, and might not always apply to this type (for interfaces for instance).

I wonder what would happen if I added the same tests you did, if this would still be passing 🤔 

https://github.com/jungsoft/rajska/blob/a5d8a0646631113b51bd693366234839559692cd/lib/middlewares/object_authorization.ex#L66

---

It doesn't ☹️ 

I believe the problem is "one layer down":
> Project one layer down from where we are right now.

In the tests we have a fragment within a fragment and so the inner fragment is ignored by the authorization and the query returns successfully when it should return an error. 

So I think the best option here is for me to merge #29 and then you can rebase with it. Do you agree?

---

@gabrielpra1 noticed it yesterday yeah, have been trying to get it to work but haven't been succesful yet. The thing is, not all fragments are named:

```gql
... on Type {
  field
}
```

I don't think this one would work, would it?

I still think projection should be the solution here, and as long as we project at the correct level, it should error appropriately, even in nested fragments.

---

@gabrielpra1 the thing for us here, is that we'd really like to be able to use the same query for users with different permissions, and those permissions would only check *actual* access. For instance, in relay, you might see queries like this:

```gql
query NodeQuery($id: ID!) {
  node(id: $id) {
    ...Product
    ...Order
    ...Customer
  }
}
```

As long as the user only receives what they have access to (e.g. only passes global IDs of types that are allowed), this is fine. The thing is, node is defined as an interface, and only *after* resolution can we be sure that the user isn't accessing anything it doesn't have access to. I naively assumed that the middleware was being created for each object returned, but it only runs once, before resolution, meaning the details required to determine what the node interface resolves to, are not available yet.

---

> I don't think this one would work, would it?

It should work, we already support inline fragments. Here is the query being used in the test case:

```
    {
      unionQuery {
        ... on User {
          name
        }
        ... on WalletBalance {
          total
        }
      }
    }
```

Absinthe's projection only takes the entire resolution as argument, so I don't think we could use it recursively. Seeing the source code, though, I think we are already covering all cases, except the `skip` directive, which I have never seen being used.

---

We do use `description @include(if: $includeProductDescription)` which is similar to `@skip`.

I can try your PR in our project to see if any issues arise.

[https://hexdocs.pm/absinthe/Absinthe.Resolution.html#project/2](The example from the docs) seem to suggest it would be able to run at any level, I think because of the `path` in the resolution, but again, this would not make it work with interface types (it would work only post-resolution)

---

I'll merge it for now since it is covering our use cases, but we can continue that discussion in another PR if you still have troubles with it. 

I'm not sure how to deal with interface types, it looks like we would need to rely only on the ObjectScopeAuthorization for this case, since it is the only one that runs after resolution